cmd/clusterctl/pkg/client/cluster/upgrader.go

```diff
@@ -0,0 +1,275 @@
+	// for reading the most recent metadata available in the repository.
+	var latestVersion *version.Version
+	for _, availableVersion := range allAvailableVersions {
+		availableSemVersion, err := version.ParseSemantic(availableVersion)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "failed to parse available version for the %s/%s provider", provider.Namespace, provider.Name)
+		}
+		if latestVersion == nil || latestVersion.LessThan(availableSemVersion) {
+			latestVersion = availableSemVersion
+		}
+	}
+
+	// Read the most recent metadata.
+	providerMetadata, err := providerRepository.Metadata(versionTag(latestVersion)).Get()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Validate consistency between metadata and versions.
+
+	// Check if the releaseSeries defined in metadata includes the current version.
+	currentSemVersion, err := version.ParseSemantic(provider.Version)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed to parse current version for the %s/%s provider", provider.Namespace, provider.Name)
+	}
+
+	if !isInReleaseSeries(currentSemVersion, providerMetadata) {
+		return nil, nil, errors.Errorf("invalid provider metadata: version %s (the current version) for the provider %s/%s does not match any release series", provider.Version, provider.Namespace, provider.Name)
+	}
+
+	// Check if the releaseSeries defined in metadata includes all the available versions for the provider (version older than the current version are dropped because not relevant for upgrade plan).
+	var availableVersions []version.Version
+	for _, availableVersion := range allAvailableVersions {
+		// we are ignoring the conversion error here because a first check already passed above
+		availableSemVersion, _ := version.ParseSemantic(availableVersion)
+
+		if currentSemVersion.LessThan(availableSemVersion) {
+			if isInReleaseSeries(availableSemVersion, providerMetadata) {
+				availableVersions = append(availableVersions, *availableSemVersion)
+				continue
+			}
+			return nil, nil, errors.Errorf("invalid provider metadata: version %s (one of the available versions) for the provider %s/%s does not match any release series", availableVersion, provider.Namespace, provider.Name)
+		}
+	}
+
+	sort.Slice(availableVersions, func(i, j int) bool {
+		return availableVersions[i].LessThan(&availableVersions[j])
+	})
+
+	return providerMetadata, availableVersions, nil
+}
+
+// getClusterAPIVersionForUpgrade return the list of ClusterAPI version available for a provider upgrade. e.g.
+// - the current version of the provider is linked with v1alpha3 ClusterAPIVersion (the latest) --> v1alpha3
+// - the current version of the provider is linked with v1alpha3 ClusterAPIVersion but there is also v1alpha4 ClusterAPIVersion defined --> v1alpha3, v1alpha4
+func getClusterAPIVersionForUpgrade(currentVersion version.Version, metadata *clusterctlv1.Metadata) []string {
```

Comment:
The name of this function is a little misleading, when I first saw the call, I assumed it'd return only one, then only to find out it returns a slice.